File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        helper(root, result, 0);
        return result;
    }
    
    void helper(TreeNode root, List<Integer> result, int i) {
        if (root == null) return;
        
        if (result.size() == i) result.add(root.val);
        
        helper(root.right, result, i + 1);
        helper(root.left, result, i + 1);
    }
}

File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
import scala.collection.mutable._

object Solution {
  def rightSideView(root: TreeNode): List[Int] = {
    val res: ListBuffer[Int] = new ListBuffer()
    helper(root, 0, res)
    res.toList
  }

  def helper(root: TreeNode, d: Int, res: ListBuffer[Int]): Unit = {
    if (root != null) {
      if (d >= res.length) res += root.value
      helper(root.right, d + 1, res)
      helper(root.left, d + 1, res)
    }
  }
}